src/api.v2/helpers/pipeline/gem2s.js

```diff
@@ -0,0 +1,173 @@
+
+  io.sockets.emit(`ExperimentUpdates-${experimentId}`, response);
+};
+
+const generateGem2sParams = async (experimentId, authJWT) => {
+  const defaultMetadataValue = 'N.A.';
+
+  logger.log('Generating gem2s params');
+
+  const getS3Paths = (files) => (
+    {
+      matrix10x: files.matrix10x.s3Path,
+      barcodes10x: files.barcodes10x.s3Path,
+      features10x: files.features10x.s3Path,
+    }
+  );
+
+  const [experiment, samples] = await Promise.all([
+    new Experiment().findById(experimentId).first(),
+    new Sample().getSamples(experimentId),
+  ]);
+
+  const samplesInOrder = experiment.samplesOrder.map(
+    (sampleId) => _.find(samples, { id: sampleId }),
+  );
+
+  const s3Paths = _.map(samplesInOrder, 'files').map(getS3Paths);
+
+  const taskParams = {
+    projectId: experimentId,
+    experimentName: experiment.name,
+    organism: null,
+    input: { type: samples[0].sampleTechnology },
+    sampleIds: experiment.samplesOrder,
+    sampleNames: _.map(samplesInOrder, 'name'),
+    sampleS3Paths: s3Paths,
+    apiVersion: 'v2',
+    authJWT,
+  };
+
+  const metadataKeys = Object.keys(samples[0].metadata);
+
+  if (metadataKeys.length) {
+    logger.log('Adding metadatakeys to task params');
+
+    taskParams.metadata = metadataKeys.reduce((acc, key) => {
+      // Make sure the key does not contain '-' as it will cause failure in GEM2S
+      const sanitizedKey = key.replace(/-+/g, '_');
+
+      acc[sanitizedKey] = Object.values(samplesInOrder).map(
+        (sampleValue) => sampleValue.metadata[key] || defaultMetadataValue,
+      );
+
+      return acc;
+    }, {});
+  }
+
+  logger.log('Task params generated');
+
+  return taskParams;
+};
+
+const gem2sCreate = async (experimentId, body, authJWT) => {
+  logger.log('Creating GEM2S params...');
+  const { paramsHash } = body;
+
+  const taskParams = await generateGem2sParams(experimentId, authJWT);
+
+  const { stateMachineArn, executionArn } = await createGem2SPipeline(experimentId, taskParams);
+
+  logger.log('GEM2S params created.');
+
+  const newExecution = {
+    params_hash: paramsHash,
+    state_machine_arn: stateMachineArn,
+    execution_arn: executionArn,
+  };
+
+  await new ExperimentExecution().upsert(
+    {
+      experiment_id: experimentId,
+      pipeline_type: 'gem2s',
+    },
+    newExecution,
+  );
+
+  logger.log('GEM2S params saved.');
+
+  return newExecution;
+};
+
+const gem2sResponse = async (io, message) => {
+  AWSXRay.getSegment().addMetadata('message', message);
+
+  // Fail hard if there was an error.
+  await validateRequest(message, 'GEM2SResponse.v1.yaml');
+
+  await hookRunner.run(message);
+
+  const { experimentId } = message;
+
+  const messageForClient = _.cloneDeep(message);
+
+  // Make sure authJWT doesn't get back to the client
+  delete messageForClient.authJWT;
+  delete messageForClient.input.authJWT;
+
+  await sendUpdateToSubscribed(experimentId, messageForClient, io);
+};
+
+module.exports = {
+  gem2sCreate,
```

Comment:
I think it'd be better to split this into two files `gem2sCreate.js` and `gem2sResponse.js`. They are here together but they are unrelated, on one hand there's `gem2Create` and it's 3 helper functions and on the other `gem2sResponse`. Just a style decision, I'm ok with keeping it like this